packages/scaffolder/src/features/remoteSources.ts

```diff
@@ -0,0 +1,267 @@
+    }
+
+    const {
+      github = undefined,
+      name = undefined,
+      url = undefined,
+    } = source.github;
+
+    if (url) {
+      const { org, repo } = parseGitHubUrl(url);
+
+      if (!org || !repo) {
+        throw new Error(`Invalid GitHub URL: ${url}`);
+      }
+
+      return `${getCheckoutBaseDirectory()}/github/${escapeNonAlphanumeric(org)}/${escapeNonAlphanumeric(repo)}`;
+    } if (name) {
+      const [org, repo] = name.split('/');
+
+      return `${getCheckoutBaseDirectory()}/github/${escapeNonAlphanumeric(org)}/${escapeNonAlphanumeric(repo)}`;
+    } if (github) {
+      const [org, repo] = github.split('/');
+
+      return `${getCheckoutBaseDirectory()}/github/${escapeNonAlphanumeric(org)}/${escapeNonAlphanumeric(repo)}`;
+    }
+
+    throw new Error(`Invalid GitHub source: ${JSON.stringify(source.github)}`);
+  }
+
+  if ('git' in source) {
+    let url: string;
+
+    if (typeof source.git === 'string') {
+      url = source.git;
+    } else {
+      url = source.git.git || source.git.url || '';
+    }
+
+    if (!url) {
+      throw new Error(`Invalid Git source: ${JSON.stringify(source.git)}`);
+    }
+
+    const [, protocol, host, org, repo] = url.match(/^(git@|https:\/\/)([A-Za-z0-9_.-]*)[/|:]([A-Za-z0-9_.-]*)\/([A-Za-z0-9_.-]*)(?:\.git)?(?:#[A-Za-z0-9_.-/]*)?$/) || [];
+
+    if (!protocol || !host || !org || !repo) {
+      throw new Error(`Invalid Git URL: ${url}`);
+    }
+
+    return `${getCheckoutBaseDirectory()}/git/${escapeNonAlphanumeric(host)}/${escapeNonAlphanumeric(org)}/${escapeNonAlphanumeric(repo)}`;
+  }
+
+  throw new Error(`Unsupported source type: ${JSON.stringify(source)}`);
+}
+
+/**
+ * Ensure a local repository is up to date.
+ * Only update the local repository once every hour.
+ *
+ * @todo Allow this cache to be overridden.
+ */
+async function updateLocalRepository(source: GitSource, directory: string) {
+  const { git: cloneUrl } = source;
+
+  let cleanUrl: string;
+  let revision: string | undefined;
+
+  if (typeof cloneUrl === 'string') {
+    // Extract the branch/commit from the clone URL.
+    [cleanUrl, revision] = cloneUrl.split('#', 2);
+  } else {
+    cleanUrl = cloneUrl.url || cloneUrl.git || '';
+    revision = cloneUrl.ref || undefined;
+  }
+
+  const git = createGit(directory);
+
+  // Clean the repository of any changes or untracked files.
+  await git.clean('fd');
+
+  // Check if the repository is tracking the proper revision
+  if (revision) {
+    const { current } = await git.status();
+
+    if (current !== revision) {
+      logger().debug(`Checking out ${chalk.green(cleanUrl)} [${chalk.yellow(revision || 'default branch')}]`);
+
+      await git.checkout(revision);
+    }
+  }
+
+  // Check if the repository should be updated and no revision was specified.
+  if (fs.statSync(directory).mtimeMs > Date.now() - 3600000) {
```

Comment:
I prefer if config values (like this timeout for checking for updates) are defined in a config file in the package rather than being hardcoded into methods. It makes the configuration clearer, easier to update, and the same value can be used in multiple places, if needed.